exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+)
+
+const (
+	prefixSplitter = "+"
+)
+
+type metricID *export.Descriptor
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	sync.RWMutex
+	registerer prometheus.Registerer
+	gatherer   prometheus.Gatherer
+
+	counters map[metricID]prometheus.Counter
+	gauges   map[metricID]prometheus.Gauge
+
+	onRegisterError func(e error)
+	counterVecs     map[string]*prometheus.CounterVec
+	gaugeVecs       map[string]*prometheus.GaugeVec
+}
+
+var _ export.Exporter = (*Exporter)(nil)
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registerer is the prometheus registerer to register
+	// metrics with. Use nil to specify the default registerer.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with. Use nil to specify the default gatherer.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// DefaultSummaryObjectives is the default summary objectives
+	// to use. Use nil to specify the default summary objectives.
+	DefaultSummaryObjectives map[float64]float64
+
+	// OnRegisterError defines a method to call to when registering
+	// a metric with the registerer fails. Use nil to specify
+	// to panic by default when registering fails.
+	OnRegisterError func(err error)
```

Comment:
This is fine, but I've been thinking about how to handle constructing metric instruments at the `Meter` level, which has to deal with duplicates as well. The practice in go of prefixing the method with `Must` seems appropriate--it can panic. Otherwise, I think I'd prefer to see errors propagate up to the `Meter` which has an error handler function itself. If the `Meter` is doing its job, we shouldn't really expect errors by the time we get to the exporter.

Reply:
As pointed by @shreyassrivatsan, he added this when the `Export` method didn't have an error return, I've just removed this and returned the error to `Meter`.